client/state/data-layer/wpcom-http/test/utils.js

```diff
@@ -7,124 +7,205 @@ import { spy } from 'sinon';
+
+			expect( initiator ).to.not.have.been.called;
+			expect( onSuccess ).to.not.have.been.called;
+			expect( onFailure ).to.not.have.been.called;
+			expect( onProgress ).to.have.been.calledWith( store, progress, progressInfo );
+		} );
+
+		it( 'should not throw runtime error if onProgress is not specified', () => {
+			dispatcher = dispatchRequest( initiator, onSuccess, onFailure );
+			expect( () => dispatcher( store, progress ) ).to.not.throw( TypeError );
+		} );
+
+		it( 'should validate response data', () => {
+			const schema = {
+				type: 'object',
+				properties: { count: { type: 'integer' } },
+			};
+
+			const fromApi = makeParser( schema );
+			dispatchRequest( initiator, onSuccess, onFailure, { fromApi } )( store, success, data );
+
+			expect( initiator ).to.not.have.been.called;
+			expect( onSuccess ).to.have.been.called;
+			expect( onFailure ).to.not.have.been.called;
+		} );
+
+		it( 'should fail-over on invalid response data', () => {
+			const schema = {
+				type: 'object',
+				properties: { count: { type: 'string' } },
+			};
+
+			const fromApi = makeParser( schema );
+			dispatchRequest( initiator, onSuccess, onFailure, { fromApi } )( store, success, data );
+
+			expect( initiator ).to.not.have.been.called;
+			expect( onSuccess ).to.not.have.been.called;
+			expect( onFailure ).to.have.been.called;
+		} );
+
+		it( 'should validate with additional fields', () => {
+			const schema = {
+				type: 'object',
+				properties: { count: { type: 'integer' } },
+			};
+
+			const extra = { count: 15, is_active: true };
+			const action = { type: 'REFILL', meta: { dataLayer: { data: extra } } };
+
+			const fromApi = makeParser( schema );
+			dispatchRequest( initiator, onSuccess, onFailure, { fromApi } )( store, action, extra );
+
+			expect( initiator ).to.not.have.been.called;
+			expect( onSuccess ).to.have.been.called;
+			expect( onFailure ).to.not.have.been.called;
+		} );
+
+		it( 'should filter out additional fields', () => {
+			const schema = {
+				type: 'object',
+				properties: { count: { type: 'integer' } },
+			};
+
+			const extra = { count: 15, is_active: true };
+			const action = { type: 'REFILL', meta: { dataLayer: { data: extra } } };
+
+			const fromApi = makeParser( schema );
+			dispatchRequest( initiator, onSuccess, onFailure, { fromApi } )( store, action, extra );
+
+			expect( initiator ).to.not.have.been.called;
+			expect( onSuccess ).to.have.been.called;
+			expect( onSuccess ).to.have.been.calledWithExactly( store, action, { count: 15 } );
+			expect( onFailure ).to.not.have.been.called;
+		} );
+
+		it( 'should transform validated output', () => {
+			const schema = {
+				type: 'object',
+				properties: { count: { type: 'integer' } },
+			};
+
+			const extra = { count: 15, is_active: true };
+			const action = { type: 'REFILL', meta: { dataLayer: { data: extra } } };
+
+			const transformer = ( { count } ) => ( { tribbleCount: count * 2, haveTrouble: true } );
```

Comment:
Wow. Had to search to get the tribble trouble reference.

Reply:
<img src="http://www.ex-astris-scientia.org/database/tribbles/tribble-thetroublewithtribbles2.jpg" width="300" />